Mixtral: Mixture of Experts quantization

casper-hansen opened the pull request:
BIG NOTE: Pending more perplexity numbers. Looking to see if we can optimize before merging.

- FP16: Perplexity 4.137
- First iteration: Perplexity 6.469 [581b416](https://github.com/casper-hansen/AutoAWQ/pull/251/commits/581b4161f106a3c6dd43aa6ad6a1cdefa255cffc)
- Second iteration: Perplexity 5.165 [240fdc8](https://github.com/casper-hansen/AutoAWQ/pull/251/commits/240fdc879e265955b6f1e14bff50067427829e69)
- Third iteration: Perplexity 4.294 [075822c](https://github.com/casper-hansen/AutoAWQ/pull/251/commits/075822c5eb3e243ecdd95bbdbc7a8852e0339f8d)

L1aoXingyu commented on awq/models/mixtral.py:
```diff
@@ -0,0 +1,139 @@
+            expert_index = i+1
+
+            layers.append(dict(
+                prev_op=expert.w3,
+                layers=[expert.w1, expert.w3],
```
why `prev_op` and `layers` all contain `expert.w3`

L1aoXingyu replied:
it means don't apply scaling for the remaining experts

casper-hansen replied:
Feel free to test out other ways of quantizing Mixtral and make a contribution. I am in the process of figuring out how it can be scaled without losing much in perplexity.

L1aoXingyu commented on awq/models/mixtral.py:
```diff
@@ -0,0 +1,139 @@
+                    module.block_sparse_moe.experts[0].w3],
+            inp=input_feat[f'block_sparse_moe.experts.{0}.w1'],
+            module2inspect=module.block_sparse_moe.experts[0],
+        ))
```
it seems first expert missing `w2`

casper-hansen replied:
Yes, I'm aware. It doesn't improve quantization error though.

noah-kim-theori commented:
Referring to your code, I implemented Mixtral (transformers==4.36.2) in AutoAWQ 0.1.7 as a single file [gist](https://gist.github.com/noah-kim-theori/4fc23189d22d2e5ea49ef127cdb5cdfd). The all FFN of the MoE has `post_attention_layernorm` as the previous operation at [transformers.models.mixtral.modeling_mixtral#L748](https://github.com/huggingface/transformers/blob/v4.36.2/src/transformers/models/mixtral/modeling_mixtral.py#L748). In your code, is there a specific reason why the remaining expert has a previous operation as the last ffn layer at the previous expert?

L1aoXingyu commented:
> Referring to your code, I implemented Mixtral (transformers==4.36.2) in AutoAWQ 0.1.7 as a single file [gist](https://gist.github.com/noah-kim-theori/4fc23189d22d2e5ea49ef127cdb5cdfd). The all FFN of the MoE has `post_attention_layernorm` as the previous operation at [transformers.models.mixtral.modeling_mixtral#L748](https://github.com/huggingface/transformers/blob/v4.36.2/src/transformers/models/mixtral/modeling_mixtral.py#L748). In your code, is there a specific reason why the remaining expert has a previous operation as the last ffn layer at the previous expert?

I have the similar implementation and got the ppl result like this

| model | wikitext-2 | ptb | c4 |  
|:--|:--|:--|:--|
| fp16 | 4.135127067565918 | 15.207035064697266 | 8.126235008239746 |  
| int4-rtn | 4.332545280456543 | 15.017653465270996 | 8.374772071838379 |  
| int4-awq | 4.277602672576904 | 15.137899398803711 | 8.237201690673828 |

To my surprise, just using rtn can get a very strong performance.

noah-kim-theori commented:
I checked right before, and I actually think that using RTN alone produces better results too. Thanks for your share.

casper-hansen commented:
> Referring to your code, I implemented Mixtral (transformers==4.36.2) in AutoAWQ 0.1.7 as a single file [gist](https://gist.github.com/noah-kim-theori/4fc23189d22d2e5ea49ef127cdb5cdfd). The all FFN of the MoE has `post_attention_layernorm` as the previous operation at [transformers.models.mixtral.modeling_mixtral#L748](https://github.com/huggingface/transformers/blob/v4.36.2/src/transformers/models/mixtral/modeling_mixtral.py#L748). In your code, is there a specific reason why the remaining expert has a previous operation as the last ffn layer at the previous expert?

Thanks for a reference implementation. I have been exhausting GPU credits trying to scale this model effectively. There is no specific reason other.

Do you want to raise a PR to merge your changes into this branch/PR so we can merge it into AutoAWQ? I can also do it if you don’t mind.

casper-hansen commented:
> > Referring to your code, I implemented Mixtral (transformers==4.36.2) in AutoAWQ 0.1.7 as a single file [gist](https://gist.github.com/noah-kim-theori/4fc23189d22d2e5ea49ef127cdb5cdfd). The all FFN of the MoE has `post_attention_layernorm` as the previous operation at [transformers.models.mixtral.modeling_mixtral#L748](https://github.com/huggingface/transformers/blob/v4.36.2/src/transformers/models/mixtral/modeling_mixtral.py#L748). In your code, is there a specific reason why the remaining expert has a previous operation as the last ffn layer at the previous expert?
> 
> I have the similar implementation and got the ppl result like this
> 
> model	wikitext-2	ptb	c4
> fp16	4.135127067565918	15.207035064697266	8.126235008239746
> int4-rtn	4.332545280456543	15.017653465270996	8.374772071838379
> int4-awq	4.277602672576904	15.137899398803711	8.237201690673828
> To my surprise, just using rtn can get a very strong performance.

I updated the code with the new quantization of layers, I got Perplexity 4.294. What did you do differently from the current implementation?

noah-kim-theori commented:
I checked your commit, and it's fine to use it as is. Feel free to use it.

vince62s commented:
Did you guys run a MMLU benchmark on the quantized model? I'm a bit disappointed. getting 60 vs 71
did not have such a discrepancy with the mistral instruct.

casper-hansen commented:
> Did you guys run a MMLU benchmark on the quantized model? I'm a bit disappointed. getting 60 vs 71 did not have such a discrepancy with the mistral instruct.

Did you evaluate with fused modules?

vince62s commented:
Well I'm using OpenNMT-py but I benchmarked (speed-wise) your code and in fact the only 2 big things are fasttransformer (that I replaced by flash2 with kv cache doing the same stuff) and "your" RMSnorm kernel, both of them making the nice speed. btw gemv works fine for batches > 1, just a little slower than gemm but works ok.

casper-hansen commented:
> Well I'm using OpenNMT-py but I benchmarked (speed-wise) your code and in fact the only 2 big things are fasttransformer (that I replaced by flash2 with kv cache doing the same stuff) and "your" RMSnorm kernel, both of them making the nice speed. btw gemv works fine for batches > 1, just a little slower than gemm but works ok.

Nice, I have been looking to replace FasterTransformer modules with Flash Attention. The kernels that are in AutoAWQ are imported from other projects to maximize inference speed and to create generalized modules. GEMV is great in many cases, especially for local models!

vince62s commented:
False alarm, I am getting 67.1 using the right Rope Theta. btw don't forget to make it an option bc @younesbelkada is already tagging this PR :)

younesbelkada commented:
Thanks @vince62s you mean in the transformers integration for fused modules?

vince62s commented:
yes here: https://github.com/casper-hansen/AutoAWQ/blob/main/awq/modules/fused/attn.py#L224
@casper-hansen knows because he mentioned this in an issue but in case some people already tries this PR.

younesbelkada commented:
Ah yes makes sense, thanks for the heads, will update once I raise the PR in transformers!

exceedzhang commented:
I ran Mixtral8*7b-v0.1 model.quantize error!
File "/data1/apps/miniconda3/envs/Mixtral/lib/python3.10/site-packages/awq/modules/linear.py", line 79, in from_linear
    qweight[:, col] |= qweight_col << (i * awq_linear.w_bit)
IndexError: index 0 is out of bounds for dimension 1 with size 0
<img width="1046" alt="image" src="https://github.com/casper-hansen/AutoAWQ/assets/4583537/3374fec0-b7fc-4061-9b53-456d958c7971">

<img width="1046" alt="image" src="https://github.com/casper-hansen/AutoAWQ/assets/4583537/1fdbc0ce-e62b-4bec-8942-0b477e8b9743">

vince62s commented:
for the sake of completeness, I ran my same mmlu script on the HF model from @casper-hansen 
ACC-all: 0.6682
So the calibration impact on PPL is clear but the impact on the MMLU benchmark is nil. I am wondering whether the actual output is better or not after calibration.

vince62s commented:
I am using my own adaptation (for OpenNMT-py) of this script https://github.com/FranxYao/chain-of-thought-hub/tree/main/MMLU which is almost the original implementation of the MMLU (slightly different of lm_eval harness used by HF leader board).
In MMLU you expect one token being A, B, C, D. If the output is "Doe" instead of [A, B, C, D] then wrong answer, HF Leader board will take the best score out of [A, B, C, D] so there is always an answer. Anyway just a slight difference and the above script is way faster.
